### PR TITLE
12.0 mig auth oauth multi token

### PR DESCRIPTION
**Pull request by fmdl**
@sla-subteno-it
@gurneyalex

### CODE REVIEW

File: auth_oauth_multi_token/README.rst [hunk 1]
```diff
@@ -0,0 +1,64 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=================
+OAuth multi token
+=================
+
```
Comment (tbaden) [inline auth_oauth_multi_token/README.rst:8]: please use the new fragment [readme](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0)

File: auth_oauth_multi_token/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+{
+    'name': 'OAuth Multi Token',
+    'version': '12.0.1.0.0',
+    'license': 'AGPL-3',
+    'author': 'Florent de Labarre, '
+              'Camptocamp, '
+              'Odoo Community Association (OCA)',
+    'summary': """Allow multiple connection with the same OAuth account""",
+    'category': 'Tool',
+    'website': 'https://github.com/OCA/server-tools',
```
Comment (tbaden) [inline auth_oauth_multi_token/__manifest__.py]: url should be https://github.com/OCA/server-auth/

File: auth_oauth_multi_token/views/res_users.xml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="user_oauth_multi_token_form" model="ir.ui.view">
```
Comment (tbaden) [inline auth_oauth_multi_token/views/res_users.xml]: optional change:
use original id (view_users_form)

File: auth_oauth_multi_token/models/res_users.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    )
+
+    @property
+    def multi_token_model(self):
+        return self.env['auth.oauth.multi.token']
+
+    @api.model
+    def _auth_oauth_signin(self, provider, validation, params):
+        """Override to handle sign-in with multi token."""
+        res = super(ResUsers, self)._auth_oauth_signin(
```
Comment (tbaden) [inline auth_oauth_multi_token/models/res_users.py]: you can shorten it to res = super()._auth_oauth_signin(...)

File: auth_oauth_multi_token/models/res_users.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        """Inactivate current user tokens."""
+        self.mapped('oauth_access_token_ids')._oauth_clear_token()
+        for res in self:
+            res.oauth_master_uuid = self._generate_oauth_master_uuid()
+
+    @api.model
+    def _check_credentials(self, password):
+        """Override to check credentials against multi tokens."""
+        try:
+            return super(ResUsers, self)._check_credentials(password)
```
Comment (tbaden) [inline auth_oauth_multi_token/models/res_users.py]: you can shorten this to:
return super()._check_credentials(password)

File: auth_oauth_multi_token/models/res_users.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            res = self.multi_token_model.sudo().search([
+                ('user_id', '=', self.env.uid),
+                ('oauth_access_token', '=', password),
+                ('active_token', '=', True),
+            ])
+            if not res:
+                raise
+
+    def _get_session_token_fields(self):
+        res = super(ResUsers, self)._get_session_token_fields()
```
Comment (tbaden) [inline auth_oauth_multi_token/models/res_users.py]: you can shorten it to:
res = super()._get_session_token_fields()

File: auth_oauth_multi_token/tests/test_multi_token.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class TestMultiToken(SavepointCase):
+
+    post_install = True
+    at_install = False
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestMultiToken, cls).setUpClass()
```
Comment (tbaden) [inline auth_oauth_multi_token/tests/test_multi_token.py]: you can shorten it to:
super().setUpClass()

File: auth_oauth_multi_token/models/res_users.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    def _generate_oauth_master_uuid(self):
+        return uuid.uuid4().hex
+
+    oauth_access_token_ids = fields.One2many(
+        comodel_name='auth.oauth.multi.token',
+        inverse_name='user_id',
+        string='OAuth tokens',
+        copy=False,
+        groups='base.group_system'
```
Comment (tbaden) [inline auth_oauth_multi_token/models/res_users.py:25]: what's the specific reason for adding the group now?

**Review by tbaden — CHANGES_REQUESTED** [review]
thanks for the work 👍 
I did a code review:
here some suggested changes

File: auth_oauth_multi_token/models/res_users.py (hunk 8 above)
Comment (fmdl) [reply auth_oauth_multi_token/models/res_users.py:25]: a standard user have no access to `auth.oauth.multi.token`

File: auth_oauth_multi_token/README.rst (hunk 1 above)
Comment (fmdl) [reply auth_oauth_multi_token/README.rst:8]: Predo says is not mandatory.

**Comment by fmdl** [comment]
@tbaden updated.

**Review by sla-subteno-it — APPROVED** [review]

**Comment by fmdl** [comment]
cc @simahawk

File: auth_oauth_multi_token/README.rst (hunk 1 above)
Comment (tbaden) [reply auth_oauth_multi_token/README.rst:8]: @fmdl are you sure? It's in the migration guidelines. You dont have to create the readme, as it's generated from the fragment rst files, but you have no fragment files.

File: auth_oauth_multi_token/views/auth_oauth_multi_token.xml [hunk 9]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="view_auth_oauth_multi_token_form" model="ir.ui.view">
+        <field name="name">auth_oauth_multi_token form</field>
+        <field name="model">auth.oauth.multi.token</field>
+        <field name="type">form</field>
+        <field name="arch" type="xml">
+            <form>
+                <sheet>
+                    <group>
```
Comment (simahawk) [inline auth_oauth_multi_token/views/auth_oauth_multi_token.xml]: name the group

**Review by simahawk — APPROVED** [review]
LG

File: auth_oauth_multi_token/views/auth_oauth_multi_token.xml (hunk 9 above)
Comment (fmdl) [reply auth_oauth_multi_token/views/auth_oauth_multi_token.xml]: @simahawk done

**Comment by fmdl** [comment]
@pedrobaeza could you merge, thanks
